### ISSUE TITLE
switch  was a thing on PCs so still uses the legacy DirectInput API.

### ISSUE THREAD

**Issue by PatrickvL** [Cxbx-Reloaded/Cxbx-Reloaded#156]

**Comment by LavenderMoon** [comment]
You almost never want to switch away from DirectInput to XInput, as that leaves a bunch of old controllers in the dust (such as my original Xbox controller, connected to PC via some cheap GCN/PS2/Xbox to USB converter). Generally, you want to implement [both at the same time](https://msdn.microsoft.com/en-us/library/windows/desktop/ee417014%28v=vs.85%29.aspx#XInput_and_DirectInput_Side_by_Side).

Additionally, FNA uses SDL's Game Controller API under the hood, and that seems to work perfectly for me with both my Xbox 360 controller and my N64-knockoff USB controller. As far as I understand, it _should_ be possible to use SDL solely for one subsystem (ie. Game Controller API) while handling the rest on your own, meaning that you could just use SDL for controller input and ignore the rest.

**Comment by jarupxx** [comment]
CXBX uses the DirectInput API, but it is missing support of POV-hat input. Implementing this should detected D-Pad. (All the other buttons and axes on the controller work. #155 #28)
This approach also supports legacy controllers.

**Comment by PatrickvL** [comment]
There seem to exist multiple solutions to the dillema of how to handle game controllers.

SDL2 indeed offers an API for this, and is portable across platforms, but it also contains a lot of stuff we probably won't be using.

We prefer a input-handling library that's:
* small
* standalone
* portable
* consists of clean C(++) code
* has good documenation
* is actively maintained

If the SDL2 Joystick/GameController API is (or can be) extracted into an independent library, that could be one possible candidate. But perhaps others are a more natural fit - let's build up an inventory of candidates, discuss them and decide which one we'll use.

**Comment by x1nixmzeng** [comment]
You can of course initialize SDL to use the `SDL_INIT_GAMECONTROLLER` subsystem but I'm not about building specific modules - https://wiki.libsdl.org/SDL_InitSubSystem

**Comment by BenNottelling** [comment]
As a solution to losing support for legacy this might provide an easy solution:
www.pcgamer.com/how-to-use-old-directinput-controllers-in-new-xinput-games/
and
www.x360ce.com/

**Comment by ciaran2** [comment]
Sniffing around for libraries that meet the criteria PatrickvL laid out, gainput looks promising:

gainput - https://github.com/jkuhlmann/gainput

Cross-platform, MIT License, minimal dependencies, seems to be under reasonably active development, and all it does is input.

If it would be helpful, I could try porting the input system in a fork and see how that goes.  If nothing else, it would be a good starting point for abstracting platform-specific details out of the input interface.

**Comment by PatrickvL** [comment]
Oh yes, please do!

**Comment by PatrickvL** [comment]
@ciaran2 How's it going?

**Comment by ciaran2** [comment]
I've had limited time because I've been crunching at work, but I've written some simple programs using the gainput API, and I think I'm close enough to grokking the interface between the CXBX input subsystem and the rest of the emulator to start ripping out and replacing the DInput bits.  I'll keep you posted.

As sort of a background concern, I don't see a clear way to enumerate the currently connected devices in the gainput API so that only available devices are registered.  The current version of the emulator only support one controller anyway, though, so I don't think it's an imminent concern.  I'll just raise this with the project lead and proceed with what I'm doing.

**Comment by PatrickvL** [comment]
You could submit an issue in their GitHub about that, perhaps there's an easy fix possible or some API already exists for that you just haven't found yet.

As for how Cxbx currently works: You don't have to follow the shared memory approach - maybe a separate thread in the emulation run time environment (possibly using a separate thread to handle the window messages loop) would be simpler to implement

**Comment by LukeUsher** [comment]
I want to move away from the shared memory approach if possible, I think it's a source of bugs. ,dsadsdaflskajhfioh32owhwfhionasv

**Comment by LukeUsher** [comment]
I want to move away from shared memory wherever possible as it is a source of bugs. I'd get away from using the registry to save settings to: A config file in the appdata folder would be better,

**Comment by ciaran2** [comment]
The registry thing was definitely at the top of my list for removal from
the interface.  In fact, even file system paths should be abstracted in one
place, no?

**Comment by PatrickvL** [comment]
@ciaran2 I see you've created https://github.com/jkuhlmann/gainput/issues/33 and got an answer, I hope it helps!

**Comment by blueshogun96** [comment]
I'm still having issues with the Dpad on my end.  Keys for it never get bound, plus since the key doesn't get bound, I get sporadic movement as if the emulator is accepting bogus input from somewhere else.  So on games that actually use the dpad, everything keeps moving to the left, making it harder to actually select stuff.

Is there a solution for this?

Shogun.

**Comment by LukeUsher** [comment]
Generally, clearing the Cxbx-Reloaded registry key and re-configuring the input fixes this. Also there seems to be an issue that if a USB controller is attached, input gets confused. I plan on re-working the configuration system but it's still quite low down on the list

**Comment by ObiKKa** [comment]
My make shift solution after multiple retries over days was that I put my mouse up above the desk surface before clicking on a button to configure (to make sure the laser movement doesn't affect/distort your key press (or your gamepad's controls). I did that with some movement sets. And it worked fine for me again in Battlestar Galactica.

**Comment by LukeUsher** [comment]
If it's the mouse that's the problem, perhaps we should prevent binding to mouse for now. I understand that some users may wish to have this functionality, eg: mapping camera controllers to mouse for games that have them, but for now it's causing us issues.

**Comment by blueshogun96** [comment]
Btw, I cleared the registry key, and that didn't work.  Still can't get the dpad to work.  Makes things harder to test depending on the title :(

Shogun

**Comment by ciaran2** [comment]
Looks like there's discussion here about totally refactoring how the input subsystem talks to the rest of the emulator.  Sounds good, but it'll probably be a couple months before I have time to seriously look at a major overhaul like that.

On the gainput side, it looks like I don't have access to equipment that will let me test XInput pads like I thought I did. :( I can still try to work a little on keyboard and mouse stuff in between the work crunch I mentioned. In the long term, I think the Dolphin solution of leveraging a bunch of input libraries in parallel is probably the way to go.

**Comment by PatrickvL** [comment]
Okay thanks for sharing your thoughts about this.
To add to this I've read some emulators 'just' use SDL input, which might be worth looking into too

**Comment by LukeUsher** [comment]
Despite implementing XInput support, I'm keeping this issue open: We still need an alternative to the old controller configuration to make the emulator more user-friendly for those who do not opt for XInput (Or don't have an XInput compatible controller)

**Comment by klepp0906** [comment]
this looks old, but the issue seems to be reported sporatically and over some time.  leaves me unsure if it was ever addressed, if the issue is on my end, or if its something in the middle.

im trying to use a steam controller in which all buttons register to map EXCEPT the "d-pad"

**Comment by RadWolfie** [comment]
Quote from @LukeUsher:
> There is no issue with D-PAD emulation, it's rather an issue with controllers.

For some reason, controller's d-pad doesn't work with Cxbx or d-input directly. It's a known issue for years by the way. Plus it is low priority.

**Comment by klepp0906** [comment]
i hope you mean "low priority" relative to the overall youth of the emulator.  I imaging having a working dpad is pretty high priority if its ever to function for games in general.  

Either way, thank you for the reply - glad to know its not something on my end.  I tried everything I could think of to circumvent.   Also heard mention of them changing to xinput however thats apparently not in yet.

patience is a virtue they say ;p

**Comment by LukeUsher** [comment]
@klepp0906 XInput support exists and is working, for that you can completely ignore the Configure Controller dialog (it only deals with DirectInput) and set the controller mappings to XInput Ports, then everything just works.

If your controller has an XInput compatible mode, that would be the best approach.

**Comment by klepp0906** [comment]
ah, thats where the confusion set in then.  so the controller should work out of the box then?  Admittedly did not even try as not being able to map the controller properly mitigated the effort I was going to have to put in to get games onto my pc etc.  I know xenia functions in much the same way (i assume).  No means to map but if you boot a game, it works just fine.

Anyhow, thank you for clarifying, your time, and your work on the emulator.  truly impressive!

**Comment by ergo720** [comment]
With the [new input GUI](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/pull/1713), DirectInput is only used to handle the keyboard (and eventually the mouse), and all controllers will use SDL or XInput. So can this be considered done now?

**Comment by PatrickvL** [comment]
For keyboard (and eventually mouse) handling, `SDL_GetKeyboardState` (and eventually, `SDL_GetMouseState`) could be used to replace DirectInput with, so I /think/ it's better to keep this issue open until those two are addressed (or alternatively, create a separate issue for those) - what do you think is best?

**Comment by ergo720** [comment]
Better to keep it open until we replace DirectInput with those functions then.

**Comment by RadWolfie** [comment]
Pretty sure we have moved on to SDL and libusb input libraries for some time now. Do we still need this ticket open?

**Comment by ergo720** [comment]
It depends on if we want to replace the dinput code for the keyboard/mouse with SDL. If not, then this can be closed.

**Comment by RadWolfie** [comment]
I think SDL is perferred for keyboard/mouse. Not sure what else still uses DirectInput as they could be already in SDL's life support anyway.

**Comment by ergo720** [comment]
On second thought, I'd rather replace dinput for the kb/mo with raw input instead. This because, both dinput and SDL do not support distinguishing the input coming from multiple keyboards/mice, and will instead combine all input from these devices as a single value. This would allow us to support multiplayer lightgun gaming when using multiple kb/mo bound to the emulated lightguns, which is currently not possible, In addition, we are already using raw input internally to detect controller hotplug events.
